extensions/smb/SmbConnectionControllerService.cpp

```diff
@@ -0,0 +1,119 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "SmbConnectionControllerService.h"
+#include "core/Resource.h"
+#include "utils/OsUtils.h"
+#include "utils/expected.h"
+
+namespace org::apache::nifi::minifi::extensions::smb {
+
+void SmbConnectionControllerService::initialize() {
+  setSupportedProperties(Properties);
+}
+
+void SmbConnectionControllerService::onEnable()  {
+  std::string hostname;
+  std::string share;
+
+  if (!getProperty(Hostname, hostname))
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing hostname");
+
+  if (!getProperty(Share, share))
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing share");
+
+  server_path_ = "\\\\" + hostname + "\\" + share;
+
+  auto password = getProperty(Password);
+  auto username = getProperty(Username);
+
+  if (password.has_value() != username.has_value())
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION,  "Either both a username and a password, or neither of them should be provided.");
+
+  if (username.has_value())
+    credentials_.emplace(Credentials{.username = *username, .password = *password});
+  else
+    credentials_.reset();
+
+  ZeroMemory(&net_resource_, sizeof(net_resource_));
+  net_resource_.dwType = RESOURCETYPE_DISK;
+  net_resource_.lpLocalName = nullptr;
+  net_resource_.lpRemoteName = server_path_.data();
+  net_resource_.lpProvider = nullptr;
+}
+
+void SmbConnectionControllerService::notifyStop() {
+  auto disconnection_result = disconnect();
+  if (!disconnection_result)
+    logger_->log_error("Error while disconnecting from SMB: %s", disconnection_result.error().message());
+}
+
+std::shared_ptr<SmbConnectionControllerService> SmbConnectionControllerService::getFromProperty(const core::ProcessContext& context, const core::PropertyReference& property) {
+  std::shared_ptr<SmbConnectionControllerService> smb_connection_controller_service;
+  if (auto connection_controller_name = context.getProperty(property)) {
+    smb_connection_controller_service = std::dynamic_pointer_cast<SmbConnectionControllerService>(context.getControllerService(*connection_controller_name));
+  }
+  if (!smb_connection_controller_service) {
+    minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Missing SMB Connection Controller Service");
```

Comment:
missing `throw`

Reply:
wow good catch, would clang-tidy find this if we were to run it on windows aswell?
https://github.com/apache/nifi-minifi-cpp/pull/1634/commits/3b513dcd2120d5e5ab361e8230060721450c6551#diff-8f6654a3a37c2021e4cbe35a2c93791afb67776b83389a56a45bf78e3f32bc3eR73

Reply:
Yes, the internal clang-tidy of CLion does catch it, so it probably would.  I have added a Jira for this: https://issues.apache.org/jira/browse/MINIFICPP-2213